.github/actions/start-opensearch/action.yml

```diff
@@ -0,0 +1,199 @@
+name: 'Start OpenSearch'
+description: 'Downloads latest build of OpenSearch, installs a plugin, executes and script and then starts OpenSearch on localhost:9200'
+
+inputs:
+  platform:
+    description: 'What platform is this action running on, e.g. "windows-latest" ? '
+    required: true
+
+  opensearch-version:
+    description: 'The version of OpenSearch that should be used, e.g "3.0.0"'
+    required: true
+
+  plugin-name:
+    description: 'The the name of the plugin to use, such as security-dashboards-plugin'
+    required: true
+
+  plugin-start-script:
+    description: 'The file name for the configuration script for the plugin such as install_demo_configurations'
+    required: true
+  # plugins/opensearch-security/tools/install_demo_configuration.sh
+
+  java-version:
+    description: 'The java version to be used, must be either 11 or 17'
+    required: true
+
+  plugin-zip:
+    description: 'The name of the zip file for the plugin hosted on docker-host i.e. security-plugin.zip '
+    required: true
+
+runs:
+  using: "composite"
+  steps:
+
+    # Configure longpath names if on Windows
+    - name: Enable Longpaths if on Windows
+      if: ${{ inputs.platform == 'windows-latest' }}
+      run: git config --system core.longpaths true
+      shell: pwsh
+
+    - name: Download PaExec if on Windows
+      if: ${{ inputs.platform == 'windows-latest' }}
+      run: |
+        choco install wget --no-progress 
+        wget -q https://www.poweradmin.com/paexec/paexec.exe -P C:\Windows\System32
+      shell: pwsh
+
+    # Downloads the min version of OpenSearch Core from the latest build
+    - name: Download OpenSearch Core for Linux
+      if: ${{ inputs.platform == 'ubuntu-latest' }}
+      run: |
+        wget https://ci.opensearch.org/ci/dbc/distribution-build-opensearch/${{ inputs.opensearch-version }}/latest/linux/x64/tar/builds/opensearch/dist/opensearch-min-${{ inputs.opensearch-version }}-linux-x64.tar.gz
+        tar -xzf opensearch-*.tar.gz
+        rm -f opensearch-*.tar.gz
+      shell: bash
+
+    - name: Download OpenSearch Core for Windows
+      if: ${{ inputs.platform == 'windows-latest' }}
+      run: |
+        Invoke-WebRequest https://artifacts.opensearch.org/snapshots/core/opensearch/${{ inputs.opensearch-version }}-SNAPSHOT/opensearch-min-${{ inputs.opensearch-version }}-SNAPSHOT-windows-x64-latest.zip -Outfile opensearch-${{ inputs.opensearch-version }}.zip
+        tar -xzf opensearch-${{ inputs.opensearch-version }}.zip
+        del opensearch-${{ inputs.opensearch-version }}.zip
+        Rename-Item opensearch-* opensearch
+      shell: pwsh
```

Comment:
This can be combined into a single statement with `peternied/download-file` [1] rather than treating the platforms differently.  See also how this was done for the BWC tests [2] PR that is out for review

[1] https://github.com/peternied/download-file
[2] https://github.com/opensearch-project/security/pull/2233/files

Reply:
I am a bit hesitant to rely on actions hosted on our personal github accounts. Perhaps if it were published to the marketplace page https://github.com/marketplace?type=actions or if we were to add a package to opensearch which included a variety of different actions we wanted to reuse across the organization? I just do not know that it is best idea to set precedent for relying on actions that will break should the account no longer exist or which would allow for someone to effectively modify code without code review. I think this type of action reference would let someone (not saying they would) swap their own action code directly and then effectively change the repo's test code without any need for a PR. Since they are only changing code in their private repo I do not think a PR would be required and I think we would have to rely on people to properly enumerate their action versions in order to make the action changes require PR on the repo's side before taking affect. This is just my thoughts though and I could very well be wrong. 

 In the the meantime, I will swap the platform logic to simply install wget with chocolatey since that seems to be what you are suggesting with your own example.

Reply:
> I am a bit hesitant to rely on actions hosted on our personal github accounts.

We absolutely should be suspicious of ANY GitHub Action, even those that look like official sources.  They are effectively running code on our behalf which could change at any time.

Good news! By using a version tag, it ensures that unless the repo is deleted that action will always be referenceable.

> In the the meantime, I will swap the platform logic to simply install wget with chocolatey since that seems to be what you are suggesting with your own example.

This is fine - but it sounds like very shortly we will be in the business of maintaining GitHub actions to do many different things, I'll start a discussion around this.  The Unix model of small special purpose [1] tools speaks to me, and I think we should optimize workflows like this one around the 'core' purpose and for everything else borrow.   


[1] https://www.brandur.org/small-sharp-tools

Reply:
I definitely agree that using other pre-built actions is great when possible. It is easier for everyone and saves time and confusion. I just meant that maybe we would want to make sure that whatever actions we use are on the marketplace or merged into some to-be-made 'action' repo so that we did not have to worry about the possibility that the action code was changed without the versioning number being incremented and then the code we thought did one thing now did something else. Maybe that is not a real concern though. 

If you prefer I use the action you linked I can do that though--just let me know.

Reply:
Ask me if I'm biased without asking me if I'm biased :D - I do have a preference towards the action I made

Created an issue to talk about the logistic of shared actions.
- https://github.com/opensearch-project/.github/issues/105

Reply:
I am trying to get your action to work but haven been unable to figure out any logic other than simply calling it with an If condition for the different OSs. Perhaps you have a thought about this? But the snapshot files are named significantly differently between OSs and I have been trying for a time to get them to work any way.

Reply:
Hi @peternied , I tried for a while longer to make the code compatible with your download action testing changes on a private repo to avoid notifications (https://github.com/scrawfor99/security/tree/LoweredPermissionsInstall). That being said, I am still unable to get the changes to work without having at least as much OS-based logic as there already was. In fact, I only have Linux working now whereas both worked previously. What are your thoughts on either waiting to rebase the downloading logic until after this is merged or lending your knowledge on a good work around. I intended to build dashboard install platform for that team to use and then to supplement the security-dash-repo CI/CD process but would like to take care of this issue first.

What do you suggest and what would you like me to do?